docker/tarfile/src.go

```diff
@@ -35,14 +38,55 @@ type layerInfo struct {
 	size int64
 }
 
-// NewSource returns a tarfile.Source for the specified path.
-func NewSource(path string) *Source {
+// NewSourceFromFile returns a tarfile.Source for the specified path
+func NewSourceFromFile(path string) (*Source, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening file %q", path)
+	}
+	defer file.Close()
+
+	reader, err := gzip.NewReader(file)
+	if err != nil {
+		return &Source{
+			tarPath: path,
+		}, nil
+	}
+	defer reader.Close()
+
+	return NewSourceFromStream(reader)
+}
+
+// NewSourceFromStream returns a tarfile.Source for the specified inputStream.
+func NewSourceFromStream(inputStream io.ReadCloser) (*Source, error) {
```

Comment:
`inputStream` can be just an `io.Reader` AFAICT.

Please document that the input must be uncompressed.

It would also be nice to document that it is safe for the caller to close the stream immediately after this function returns. (Otherwise, passing a stream to a “read something” object might make the impression that the stream will be consumed over time, and somehow closed at a later time.)

Reply:
fixed